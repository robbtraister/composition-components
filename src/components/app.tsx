'use strict'

import PropTypes from 'prop-types'
import React from 'react'

import withQuarantine from './quarantine'
import appContext from '../contexts/app'
import componentContext from '../contexts/component'

const isClient = typeof process === 'undefined'

const identity = x => x

function descendants({ children }) {
  const childArray = [].concat(children || [])
  return childArray.concat(...childArray.map(descendants))
}

function App({
  node,
  promises = [],
  quarantine = isClient,
  Root = React.Fragment,
  getComponent,
  getContent
}) {
  const wrapComponent = quarantine ? withQuarantine : identity

  const componentCache = {}
  function getCachedComponent(node) {
    const { type } = node
    if (!componentCache[type]) {
      componentCache[type] = wrapComponent(getComponent(type))
    }
    return componentCache[type]
  }

  const contentCache = {}
  function getCachedContent(contentParams: ContentParams) {
    const { source, query, filter } = contentParams
    const key = JSON.stringify({ source, query, filter })
    if (!contentCache[key]) {
      const contentEntry = (contentCache[key] = {
        cached: undefined,
        fetched: getContent(contentParams).then(content => {
          contentEntry.cached = content
          return content
        })
      })
      promises.push(contentEntry.fetched)
    }
    return contentCache[key]
  }

  function Node(node) {
    const Component = getCachedComponent(node)
    const { props = {} } = node
    const { children = [] } = props

    const componentContextValue = { ...node, getContent: getCachedContent }

    return (
      <componentContext.Provider value={componentContextValue}>
        <Component {...props}>
          {[].concat(children || []).map(child => (
            <Node key={child.id} {...child} />
          ))}
        </Component>
      </componentContext.Provider>
    )
  }
  Node.propTypes = {
    children: PropTypes.node
  }

  const appContextValue = {
    tree: node,
    elements: descendants({ children: node }),
    promises,
    getContent: getCachedContent
  }

  return (
    <appContext.Provider value={appContextValue}>
      <Root>
        <Node {...node} />
      </Root>
    </appContext.Provider>
  )
}

App.propTypes = {
  node: PropTypes.shape({
    type: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    props: PropTypes.object
  }),
  promises: PropTypes.array,
  quarantine: PropTypes.bool,
  Root: PropTypes.elementType,
  getComponent: PropTypes.func,
  getContent: PropTypes.func
}

export default App
export { App }
