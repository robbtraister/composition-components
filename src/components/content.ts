'use strict'

import { useContext, useState } from 'react'

import { render } from './render'

import componentContext from '../contexts/component'
import pageContext from '../contexts/page'

export function useContent(params: Composition.ContentParams) {
  const { source, query } = params
  const key = JSON.stringify({ content: { source, query } })

  const { cache = {} } = useContext(pageContext)
  const { getContent } = useContext(componentContext)

  /**
   * cache entry will always be some combination of the following (possibly both):
   * 1. a Promise
   * 2. an Object with `data` (and possibly, `expires`) property
   */

  const { data, expires } = cache[key] || {}
  const [content, setContent] = useState(data)

  if (!(key in cache) || expires < Date.now()) {
    const contentPromise: Composition.CachedPromise = Promise.resolve(
      getContent(params)
    )

    cache[key] = Object.assign(
      contentPromise,
      // preserve cached data until update is resolved
      { data },
      // preserve cached props provided by custom getContent implementation
      contentPromise
    )

    contentPromise.then(data => {
      setContent(data)
      contentPromise.data = data
    })
  }

  return content
}

export function Content(
  props: Composition.RenderableProps<
    Composition.ContentParams,
    Composition.ContentStruct
  >
) {
  const { source, query, filter, ...passThroughProps } = props
  const content = useContent({ source, query, filter })

  return render({ ...passThroughProps, content })
}

export default Content
