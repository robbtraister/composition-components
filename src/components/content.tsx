'use strict'

import React, { useEffect, useState } from 'react'
import { useComponentContext } from '../contexts/component'

function useContent(props: ContentParams) {
  const { getContent } = useComponentContext()
  const contentResult: ContentPromise = getContent(props)
  const [content, setContent] = useState(contentResult.cached)

  useEffect(() => {
    let doUpdate = true
    const awaitFetch = async () => {
      const updated = await contentResult
      doUpdate && setContent(updated)
    }
    contentResult && awaitFetch()

    return () => {
      doUpdate = false
    }
  }, [contentResult])

  return content
}

function Content(props: ContentComponentParams) {
  const { children, component: Component, render, ...contentProps } = props
  const { source, query, filter, ...otherProps } = contentProps
  const content = useContent({ source, query, filter }) || {}

  if (Component) {
    return <Component {...otherProps} content={content} />
  } else if (render) {
    return render({ ...otherProps, content })
  } else if (children) {
    return []
      .concat(children || [])
      .map((Child, index) => (
        <Child key={index} {...otherProps} content={content} />
      ))
  }
}

export default Content
export { Content, useContent }
