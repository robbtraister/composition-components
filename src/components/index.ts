'use strict'

export * from './app'
export * from './composition'
export * from './content'
export * from './quarantine'
export * from './styles'
export * from './tree'
