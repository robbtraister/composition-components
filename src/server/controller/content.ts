'use strict'

import contentSources from '~/../build/generated/content-sources'
import request from 'request-promise-native'

import logger from '../../utils/logger'

const contentCache = {}

class ContentSource {
  name: string
  fetchDirect: Function

  constructor(name, { fetch, resolve }) {
    this.name = name

    if (fetch) {
      this.fetchDirect = fetch
    } else if (resolve) {
      this.fetchDirect = async query =>
        request(await resolve(query), { json: true })
    } else {
      throw new Error('`fetch` or `resolve` is required for content source')
    }
  }

  clear(query) {
    const cacheKey = this.getCacheKey(query)
    delete contentCache[cacheKey]
  }

  async fetch(query) {
    const cacheKey = this.getCacheKey(query)
    if (cacheKey in contentCache) {
      logger.info(`Content found in cache [${cacheKey}]`)
    } else {
      logger.info(`Content not found in cache; fetching [${cacheKey}]`)
      contentCache[cacheKey] = await this.fetchDirect(query)
    }
    return contentCache[cacheKey]
  }

  getCacheKey(query) {
    return JSON.stringify({ source: this.name, query })
  }

  async update(query) {
    this.clear(query)
    return this.fetch(query)
  }
}

const contentSourceCache = {}
export function getContentSource(source) {
  if (!(source in contentSourceCache)) {
    const contentSource = contentSources[source]
    contentSourceCache[source] = new ContentSource(source, contentSource)
  }
  return contentSourceCache[source]
}

export async function clear({ source, query }: Composition.ContentParams) {
  return getContentSource(source).clear(query)
}

export async function fetch({ source, query }: Composition.ContentParams) {
  return getContentSource(source).fetch(query)
}

export async function update({ source, query }: Composition.ContentParams) {
  return getContentSource(source).update(query)
}
