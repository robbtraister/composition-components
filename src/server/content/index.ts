'use strict'

import content from '~/build/generated/content'

export async function fetch({ source, query }) {
  return content.sources[source].fetch(query)
}
