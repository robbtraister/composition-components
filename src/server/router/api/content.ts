'use strict'

import express from 'express'

import { fetch } from '../../content'

export default function router(options) {
  const contentRouter = express()

  contentRouter.use('/fetch', async (req, res, next) => {
    res.send(await fetch(req.query))
  })

  return contentRouter
}
