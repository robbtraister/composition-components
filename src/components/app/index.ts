'use strict'

import { App as ClientApp } from './client'
import { App as ServerApp } from './server'

import { isClient } from '../../utils'

export const App = isClient ? ClientApp : ServerApp

export default App
