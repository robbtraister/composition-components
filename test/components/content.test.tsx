'use strict'

/* global describe, expect, test */

import PropTypes from 'prop-types'
import React from 'react'
import { render } from '@testing-library/react'

import App from '../../src/components/app'
import { Content, useContent } from '../../src/components/content'

function Display({ content }) {
  return <div>{content}</div>
}

Display.propTypes = {
  content: PropTypes.string
}

function ContentComponent(props) {
  return <Content {...props} component={Display} />
}

function ContentHook(props) {
  const content = useContent(props)
  return <Display content={content} />
}

function ContentRender(props) {
  return <Content {...props} render={Display} />
}

function getComponent(type) {
  switch (type) {
    case 'content-hook':
      return ContentHook
    case 'content-component':
      return ContentComponent
    case 'content-render':
      return ContentRender
  }
}

async function getContent({ source, query }) {
  return `${source}: ${query.data}`
}

const node = {
  id: 'abc',
  props: {
    source: 'source',
    query: { data: 'data' }
  }
}

async function testContentComponent(componentType) {
  const promises = []
  const result = render(
    <App
      getComponent={getComponent}
      getContent={getContent}
      node={{ ...node, type: componentType }}
      promises={promises}
    />
  )

  await Promise.all(promises)
  return result
}

describe('Content', () => {
  test('Content Component', async () => {
    const { asFragment } = await testContentComponent('content-component')
    expect(asFragment()).toMatchSnapshot()
  })

  test('Content Hook', async () => {
    const { asFragment } = await testContentComponent('content-hook')
    expect(asFragment()).toMatchSnapshot()
  })

  test('Content Render', async () => {
    const { asFragment } = await testContentComponent('content-render')
    expect(asFragment()).toMatchSnapshot()
  })
})
