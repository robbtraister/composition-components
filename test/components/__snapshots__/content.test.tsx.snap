// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Content Children 1`] = `
<DocumentFragment>
  <div>
    source: data
  </div>
</DocumentFragment>
`;

exports[`Content Content Component 1`] = `
<DocumentFragment>
  <div>
    source: data
  </div>
</DocumentFragment>
`;

exports[`Content Content Hook 1`] = `
<DocumentFragment>
  <div>
    source: data
  </div>
</DocumentFragment>
`;

exports[`Content Content Render 1`] = `
<DocumentFragment>
  <div>
    source: data
  </div>
</DocumentFragment>
`;
